File: aoc15/day_02/day_02.ml
type box = { l : int; w : int; h : int };;

let parse line =
  match List.map int_of_string (Str.split (Str.regexp "x") line) with
  | l :: w :: h :: _ -> { l; w; h }
  | _ -> { l = 0; w = 0; h = 0 }
in
let boxes = Input.parse_lines parse "input" in

let area box =
  let x = box.l * box.w in
  let y = box.l * box.h in
  let z = box.w * box.h in
  (2 * x) + (2 * y) + (2 * z) + min (min x y) z
in
let first = List.fold_left ( + ) 0 (List.map area boxes) in

let volume box = box.l * box.w * box.h in
let ribbon box =
  let x = (2 * box.l) + (2 * box.w) in
  let y = (2 * box.l) + (2 * box.h) in
  let z = (2 * box.w) + (2 * box.h) in
  min (min x y) z + volume box
in
let second = List.fold_left ( + ) 0 (List.map ribbon boxes) in

Printf.printf "First: %d\n" first;
Printf.printf "Second: %d\n" second

File: aoc15/go/day_02/day_02.go
package main

import (
	"aoc/collections"
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Box struct {
	L, W, H int
}

func create(line string) Box {
	dims := strings.Split(line, "x")
	l, _ := strconv.Atoi(dims[0])
	w, _ := strconv.Atoi(dims[1])
	h, _ := strconv.Atoi(dims[2])
	return Box{l, w, h}
}

func (b Box) areas() []int {
	return []int{2 * b.L * b.W, 2 * b.W * b.H, 2 * b.H * b.L}
}

func (b Box) perimeters() []int {
	return []int{2 * (b.L + b.W), 2 * (b.W + b.H), 2 * (b.H + b.L)}
}

func (b Box) volume() int {
	return b.L * b.W * b.H
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	first := 0
	second := 0
	for scanner.Scan() {
		box := create(scanner.Text())
		first += collections.Sum(box.areas()) + slices.Min(box.areas())/2
		second += slices.Min(box.perimeters()) + box.volume()
	}
	fmt.Println("First:", first)
	fmt.Println("Second:", second)
}
